### PR TITLE
resources: aws: ec2: Refactor CheckApply

### PR DESCRIPTION
**Pull request by jonathangold**
@purpleidea: for review once #284  is merged. I included a second commit with send/recv since it's pretty  straightforward (now that checkapply has been reorganized)

### CODE REVIEW

File: resources/aws_ec2.go [hunk 1]
```diff
@@ -687,6 +689,16 @@ func (obj *AwsEc2Res) CheckApply(apply bool) (checkOK bool, err error) {
 		return false, fmt.Errorf("instance is nil")
 	}
 
+	// store useful send/recv values when we return
+	defer func() {
```
Comment (purpleidea) [inline resources/aws_ec2.go]: Assuming instance is always set, I think this is actually a really clever way to do this! +1

File: resources/aws_ec2.go [hunk 2]
```diff
@@ -674,170 +674,141 @@ func (obj *AwsEc2Res) snsWatch() error {
 // CheckApply method for AwsEc2 resource.
 func (obj *AwsEc2Res) CheckApply(apply bool) (checkOK bool, err error) {
 	log.Printf("%s: CheckApply(%t)", obj, apply)
+	defer func() {
+		log.Printf("%s: CheckApply Done)", obj)
```
Comment (purpleidea) [inline resources/aws_ec2.go]: Typo and not needed.

**Comment by jonathangold** [comment]
@purpleidea. Removed the "CheckApply Done)" message. :P

**Comment by purpleidea** [comment]
I still see 3 commits in this PR btw

**Comment by jonathangold** [comment]
@purpleidea rebased

File: resources/aws_ec2.go [hunk 3]
```diff
@@ -684,6 +689,14 @@ func (obj *AwsEc2Res) CheckApply(apply bool) (checkOK bool, err error) {
 		return false, fmt.Errorf("instance is nil")
 	}
 
+	// store useful send/recv values when we return
+	defer func() {
+		// safely dereference the pointers (nil becomes "")
+		obj.PublicIPv4 = aws.StringValue(instance.PublicIpAddress)
+		obj.PrivateIpv4 = aws.StringValue(instance.PublicIpAddress)
```
Comment (purpleidea) [inline resources/aws_ec2.go]: I merged first commit and rebased to master. This line has two bugs in it. Hint, one of them is Ipv4, not IPv4.

**Comment by jonathangold** [comment]
@purpleidea Fixed the IP addresses. I seem to make most of my mistakes when the goal seems easy/simple. I need to train myself to be just as careful despite this tendency. :)

**Comment by purpleidea** [comment]
Sorry for the holiday delays! All merged, thank you!

**Pull request merged**